onnxruntime/core/providers/qnn/builder/opbuilder/pad_op_builder.cc

```diff
@@ -0,0 +1,247 @@
+                                     const logging::Logger& logger,
+                                     bool do_op_validation) const override ORT_MUST_USE_RESULT;
+};
+
+Status PadOpBuilder::ProcessInputs(QnnModelWrapper& qnn_model_wrapper,
+                                   const NodeUnit& node_unit,
+                                   const logging::Logger& logger,
+                                   std::vector<std::string>& input_names,
+                                   bool do_op_validation) const {
+  const auto& inputs = node_unit.Inputs();
+  // QNN Pad only has 1 input, the pads input & constant_value input need to be initializer and set as Qnn node parameter, axes input is not supported.
+  if (do_op_validation) {
+    ORT_RETURN_IF(inputs.size() > 3, "QNN Pad doesn't support axes.");
+    ORT_RETURN_IF(inputs.size() < 2, "QNN Pad requires the pads input.");
+
+    std::vector<uint32_t> input_shape;
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape of input 0.");
+    ORT_RETURN_IF(input_shape.size() > 5, "QNN Pad doesn't support more than 5 dimension");
+
+    auto& pads_input_name = inputs[1].node_arg.Name();
+    ORT_RETURN_IF_NOT(qnn_model_wrapper.IsInitializerInput(pads_input_name),
+                      "Qnn doesn't support dynamic pad input");
+    if (node_unit.Inputs().size() > 2) {
+      auto& constant_value_input_name = inputs[2].node_arg.Name();
+      ORT_RETURN_IF_NOT(qnn_model_wrapper.IsInitializerInput(constant_value_input_name),
+                        "Qnn doesn't support dynamic constant_value input");
+    }
+  }
+
+  ORT_RETURN_IF_ERROR(ProcessInput(qnn_model_wrapper, inputs[0], logger, input_names));
+
+  return Status::OK();
+}
+
+template <typename T>
+float DequantizeValue(T value, int32_t offset, float scale) {
+  return static_cast<float>(static_cast<int32_t>(value) - offset) * scale;
+}
+
+Status ProcessConstantValue(QnnModelWrapper& qnn_model_wrapper,
+                            std::vector<std::string>& param_tensor_names,
+                            const NodeUnit& node_unit,
+                            const NodeUnitIODef& input) {
+  OnnxInputInfo input_info = {};
+  ORT_RETURN_IF_ERROR(qnn_model_wrapper.GetOnnxInputInfo(input, input_info));
+  std::vector<uint8_t> unpacked_tensor;
+  // Already confirmed constant_value input is initializer in ProcessInputs()
+  ORT_RETURN_IF_ERROR(qnn_model_wrapper.UnpackInitializerData(*input_info.initializer_tensor, unpacked_tensor));
+  Qnn_Scalar_t constant_value_qnn_scalar = QNN_SCALAR_INIT;
+  // constant_value is quantized
+  if (input.quant_param.has_value()) {
+    // QNN prefers pad_constant_value quantized with quantization params same as in[0], and data stored as 32-bit signed integer
+    // Onnx doesn't gurantee it has same quantization parameter as in[0], so get back the float32 value and use non-quantized data directly
```

Comment:
Is it worth adding a check here to see if the quantization parameters for in[0] and in[1] match? If so, we can just use onnx initializer value directly.

Reply:
1. in[0] is dynamic input, constant_value (in[2] optional) is a scalar. constant_value may fill out of range of in[0] value. e.g in[0] value range [1.0f, 100.0f], constant_value = 0.0f. Unless quantization script to do the work to make a shared quantization parameter to cover them all.
2. It's a little bit complex to go that approach. Because for that case, QNN requires to use int32 value to save the data. Let's say constant_value has same quantization parameter with in[0]. Then we need to convert it to int32 from it's original type like sfixed_point8, ufixed_point8, sfixed_point16, etc.
I prefer to keep it simple to just dequantize it and directly use float32 type. QNN-onnx-converter does the same thing.

---
In reply to: [1325185282](https://github.com/microsoft/onnxruntime/pull/17508#discussion_r1325185282) [](ancestors = 1325185282)